cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -510,25 +514,51 @@ public void getReportDetails(HttpServiceRequest request, HttpServiceResponder re
       Location reportDir = reportIdDir.append(LocationName.REPORT_DIR);
       // TODO: [CDAP-13290] reports should be in avro format instead of json text;
       // TODO: [CDAP-13291] need to support reading multiple report files
-      Optional<Location> reportFile = reportDir.list().stream().filter(l -> l.getName().endsWith(".json")).findFirst();
+      Optional<Location> reportFile = reportDir.list().stream().filter(l -> l.getName().endsWith(".avro")).findFirst();
       // TODO: [CDAP-13292] use cache to store content of the reports
       // Read the report file and add lines starting from the position of offset to the result until the result reaches
       // the limit
-      if (reportFile.isPresent()) {
-        try (BufferedReader br = new BufferedReader(new InputStreamReader(reportFile.get().getInputStream(),
-                                                                          StandardCharsets.UTF_8))) {
-          String line;
-          while ((line = br.readLine()) != null) {
-            // skip lines before the offset
-            if (lineCount++ < offset) {
-              continue;
-            }
-            if (reportRecords.size() == limit) {
-              break;
-            }
-            reportRecords.add(line);
+      if (!reportFile.isPresent()) {
+        LOG.error("No report files is found under directory {}", reportDir.toURI().toString());
+        responder.sendError(500, "No report files is found for report " + reportId);
+        return;
+      }
+      try {
+        Location reportFileLocation = reportFile.get();
+        DataFileReader<GenericRecord> dataFileReader =
+            new DataFileReader<>(new File(reportFileLocation.toURI()), new GenericDatumReader<>());
+        // skip sync points until the end of file is reached or the given offset is surpassed
+        long skipLen = reportFileLocation.length() / 10;
+        long skipPoint = 0;
+        while (skipLen > 0 && dataFileReader.hasNext() && skipPoint < offset) {
```

Comment:
What is `offset`? I thought offset is by number of records, but seems like `skipPoint` is actually a byte position in the file?

Reply:
right, originally thought avro can be skipped by number of records, but it's actually by byte position. Do you think it will make sense to change the API so that the offset is actually byte position and then when returning the report records, we also return the current byte position of the avro file reader?